### PR TITLE
feat(net): Metered senders

### LINKED ISSUE #140: Add metered channel abstractions
Different parts of the codebase will exchange information via (unbounded)channels.

The benefit of an unbounded channel is that a `send` will always succeed, this is important if we want to ensure that a value will arrive as long as the channel is not closed.

Values will be buffered, however, which makes the system memory the implicit upper bound of channels. A stalled channel could cause issues here. For debugging/testing purposes we could write abstractions that keep track of how many values are being sent/received via a global counter or even via `metrics`.

### PR DESCRIPTION
**Pull request by akirillo**
Resolves #140, intended to build on top of #707

### CODE REVIEW

**Comment by akirillo** [comment]
@onbjerg would appreciate a quick glance at my current approach here to address #140!
Should I also add a `BandwidthMeter` a la #707 so that we can gauge / count outbound bandwidth from the sender?

File: crates/net/common/src/metered_sender.rs
Comment (gakonst) [inline crates/net/common/src/metered_sender.rs:24]: Wonder if this can be made generic over any kind of Sink type? @mattsse. All good if not.

**Review by gakonst — COMMENTED** [review]
Wrapping around the type and tracking makes sense, do we intend to integrate this anywhere in this PR?

File: crates/net/common/src/metered_sender.rs
Comment (onbjerg) [inline crates/net/common/src/metered_sender.rs:13]: Overall this PR looks good to me, although this highlights an issue with `metrics-derive` I think - `metered_sender` is not really a good name, it's way too generic. cc @rkrasiuk is there a way we can change this scope at the site it is used? My thinking here is that a metric like `reth.network.p2p.metered_sender.messages_sent` is not a super helpful name.

File: crates/net/common/src/metered_sender.rs
Comment (onbjerg) [inline crates/net/common/src/metered_sender.rs:17]: nit 

```suggestion
    /// Number of failed message deliveries
```

File: crates/net/common/src/metered_sender.rs
Comment (mattsse) [inline crates/net/common/src/metered_sender.rs:11]: this is not limited to network, we can use this for any bounded channel that we want to monitor.
so perhaps we move this to a separate crate?

File: crates/net/common/src/metered_sender.rs
Comment (mattsse) [inline crates/net/common/src/metered_sender.rs:19]: the problem with this is now, that we can not differentiate between two different kinds of channels.

for example, if we'd use this for channel of type `A` and channel of type `B` then we don't know which channel incremented the counters.

so I think we should use a `&'static str` approach here, instead of deriving the identifiers.

File: crates/net/common/src/metered_sender.rs
Comment (mattsse) [reply crates/net/common/src/metered_sender.rs:24]: hmm, I can't think of anything where this would be useful.

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/common/src/metered_sender.rs
Comment (onbjerg) [reply crates/net/common/src/metered_sender.rs:19]: Exactly, ref https://github.com/paradigmxyz/reth/pull/726#discussion_r1064130656

File: crates/net/common/src/metered_sender.rs
Comment (akirillo) [reply crates/net/common/src/metered_sender.rs:13]: Yeah, was a concern of mine as well - not super familiar with the metrics-derive crate but I can try playing around with some way to change the metric scope where it's used

Comment (akirillo) [reply crates/net/common/src/metered_sender.rs:13]: Also, if we want to use this in multiple places, we'd want some way to distinguish which channel's metrics we're looking at

Comment (akirillo) [reply crates/net/common/src/metered_sender.rs:13]: Oh lol all of this was mentioned in a separate comment thread my b

File: crates/net/common/src/metered_sender.rs
Comment (akirillo) [reply crates/net/common/src/metered_sender.rs]: Gotcha, just won't use metric-derive then

File: crates/net/common/src/metered_sender.rs
Comment (akirillo) [reply crates/net/common/src/metered_sender.rs:24]: sg, will keep it as-is for now and we can make the change later when there's demand for it

File: crates/net/common/src/metered_sender.rs
Comment (akirillo) [reply crates/net/common/src/metered_sender.rs]: gonna take a stab at handling #782 first to address this

**Comment by akirillo** [comment]
> Wrapping around the type and tracking makes sense, do we intend to integrate this anywhere in this PR?

Am happy to! Would appreciate any pointers on where this would be the highest signal, maybe in the `Sender`s used in the `discv4` crate?

File: crates/net/common/src/metered_sender.rs
Comment (akirillo) [reply crates/net/common/src/metered_sender.rs]: I moved it to `common/metrics/metered_sender.rs`, that work?

**Comment by onbjerg** [comment]
Marked this as blocked on #785

**Comment by akirillo** [comment]
Note to self: add docs to `metrics.md` after #785 is merged

**Comment by gakonst** [comment]
Unblocked @akirillo we got #785 in now!

**Comment by akirillo** [comment]
making use of dynamic scope now - any suggestions on where to integrate this metric? which senders do we care the most about? @gakonst

**Review by gakonst — CHANGES_REQUESTED** [review]
This probably should be under crates/metrics right?

@mattsse when u get a chance if you can point at some senders that would benefit from this

**Review by mattsse — APPROVED** [review]
lgtm

**Review by gakonst — APPROVED** [review]
looks good - we should identify where we have senders and wrap them around this in a followup

**Pull request merged**
